### PR TITLE
mongodocstore: implement unordered action lists

### PR DESCRIPTION
**Pull request by jba**
We use the BulkWrite call when possible, for performance. This affects
our ability to report errors precisely.

### CODE REVIEW

File: internal/docstore/mongodocstore/mongo.go [hunk 1]
```diff
@@ -289,49 +290,59 @@ func (c *collection) toMongoFieldPath(fp []string) string {
...
-		return err
+func sliceToLower(s []string) {
+	for i, e := range s {
+		s[i] = strings.ToLower(e)
 	}
-	mdoc, err := c.encodeDoc(a.Doc, id)
+}
+
+func (c *collection) create(ctx context.Context, a *driver.Action) error {
+	// See https://docs.mongodb.com/manual/reference/method/db.collection.insertOne.
```
Comment (shantuo) [inline internal/docstore/mongodocstore/mongo.go]: This and L309 might better be summarized and put on top of `create`.

File: internal/docstore/mongodocstore/mongo.go [hunk 2]
```diff
@@ -289,49 +290,59 @@ func (c *collection) toMongoFieldPath(fp []string) string {
...
+	return nil
 }
 
-func (c *collection) replace(ctx context.Context, a *driver.Action, upsert bool) error {
-	// See https://docs.mongodb.com/manual/reference/method/db.collection.replaceOne
+func (c *collection) prepareCreate(a *driver.Action) (mdoc, createdID interface{}, err error) {
 	id, err := c.docID(a.Doc)
+	if id == "" {
+		if c.idField == "" {
+			// If the user provides a function to get the ID rather than a field, then Create
```
Comment (shantuo) [inline internal/docstore/mongodocstore/mongo.go]: Where is `idFunc` used now?

File: internal/docstore/mongodocstore/mongo.go [hunk 3]
```diff
@@ -454,6 +488,166 @@ func (c *collection) docID(doc driver.Document) (interface{}, error) {
...
+		for i, w := range bwe.WriteErrors {
+			alerr = append(alerr, indexedError{-i - 1, gcerr.Newf(translateMongoCode(w.Code), nil, "%s", w.Message)})
+		}
+		return alerr
+	}
+	if res.DeletedCount != nDeletes {
+		alerr = append(alerr, indexedError{-1, gcerr.Newf(gcerr.NotFound, nil, "some delete failed")})
+	}
+	if res.MatchedCount != nMatches {
+		alerr = append(alerr, indexedError{-2, gcerr.Newf(gcerr.NotFound, nil, "some replace failed")})
```
Comment (shantuo) [inline internal/docstore/mongodocstore/mongo.go]: Should this `-2` be `-1`?

File: internal/docstore/mongodocstore/mongo.go [hunk 4]
```diff
@@ -480,8 +670,15 @@ func (c *collection) ErrorCode(err error) gcerrors.ErrorCode {
...
 }
 
-func sliceToLower(s []string) {
-	for i, e := range s {
-		s[i] = strings.ToLower(e)
+// Error code for a write error when no documents match a filter.
+// (The Go mongo driver doesn't define an exported constant for this.)
+const mongoNotFoundCode = 11000
+
+func translateMongoCode(code int) gcerrors.ErrorCode {
```
Comment (shantuo) [inline internal/docstore/mongodocstore/mongo.go]: Can this be merged into `ErrorCode`?

File: internal/docstore/mongodocstore/mongo.go [hunk 5]
```diff
@@ -454,6 +488,166 @@ func (c *collection) docID(doc driver.Document) (interface{}, error) {
...
+		go func() {
+			err := c.get(ctx, g.action)
+			getc <- indexedError{g.index, err}
+		}()
+	}
+
+	// Run all writes in a single BulkWrite RPC.
+	alerrBulk := c.unorderedBulkWrite(ctx, writes)
+	if alerrBulk != nil {
+		alerr = append(alerr, alerrBulk...)
```
Comment (shantuo) [inline internal/docstore/mongodocstore/mongo.go]: How about doing something similar to the firestore one, make a slice here and give a proper length to the error list?

File: internal/docstore/mongodocstore/mongo.go (hunk 4 above)
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: Done.

File: internal/docstore/mongodocstore/mongo.go (hunk 3 above)
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: I guess it doesn't matter, they could all be -1.
Done.

File: internal/docstore/mongodocstore/mongo.go (hunk 5 above)
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: The problem is how to handle the errors that aren't associated with any action (the ones with -1).

File: internal/docstore/mongodocstore/mongo.go (hunk 2 above)
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: It's called in `docID`.

File: internal/docstore/mongodocstore/mongo.go [hunk 6]
```diff
@@ -454,6 +488,166 @@ func (c *collection) docID(doc driver.Document) (interface{}, error) {
...
+		}
+		// The returned indexes of the WriteErrors are wrong. See https://jira.mongodb.org/browse/GODRIVER-1028.
+		// Until it's fixed, use negative values for the indexes in the errors we return.
+		for _, w := range bwe.WriteErrors {
+			alerr = append(alerr, indexedError{-1, gcerr.Newf(translateMongoCode(w.Code), nil, "%s", w.Message)})
+		}
+		return alerr
+	}
+	if res.DeletedCount != nDeletes {
+		alerr = append(alerr, indexedError{-1, gcerr.Newf(gcerr.NotFound, nil, "some delete failed")})
```
Comment (shantuo) [inline internal/docstore/mongodocstore/mongo.go]: If user redo the same deletes when this happen, will Mongo succeed for the remaining ones and silently ignore those that are already deleted? If so we can suggest the user to retry it.

File: internal/docstore/mongodocstore/mongo.go (hunk 1 above)
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: Moved this line above the function declaration.

File: internal/docstore/mongodocstore/mongo.go [hunk 7]
```diff
@@ -289,49 +290,59 @@ func (c *collection) toMongoFieldPath(fp []string) string {
...
-func (c *collection) replace(ctx context.Context, a *driver.Action, upsert bool) error {
-	// See https://docs.mongodb.com/manual/reference/method/db.collection.replaceOne
+func (c *collection) prepareCreate(a *driver.Action) (mdoc, createdID interface{}, err error) {
 	id, err := c.docID(a.Doc)
+	if id == "" {
+		if c.idField == "" {
+			// If the user provides a function to get the ID rather than a field, then Create
+			// can't make a unique ID for a document that's missing one, because it doesn't
+			// know how to create it or where to store it in the original document.
+			return nil, nil, err
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: Done.

File: internal/docstore/mongodocstore/mongo.go (hunk 6 above)
Comment (jba) [reply internal/docstore/mongodocstore/mongo.go]: It's much more likely that the documents just don't exist. If there were a real error, like a network issue, then I'm pretty sure the call would return an error.

**Pull request merged**